azavea/raster-vision
# Add GeoDataset to allow reading directly from source imagery during training without chipping + refactor existing datasets

## Description

### Overview
This PR adds a new type of Dataset, `GeoDataset`, that allows the Learner to read rasters and labels directly from a `Scene` during training. `GeoDataset` is further subclassed into `SlidingWindowGeoDataset` and `RandomWindowGeoDataset`, that represent two different reading methods.

These datasets can be configured using `GeoDataConfig`, which is subclassed from `DataConfig`, and accepts a `rastervision.core.data.DatasetConfig` and either a single `GeoDataWindowConfig` Config that applies to all scenes or a `Dict[scene.id --> GeoDataWindowConfig]`.

### The Dataset classes
`pytorch_learner/dataset/*.py`

The Dataset classes are now structured as shown below. (Only classificaiton and semantic segmentation have been shown, but it's the same for object detection and regression too). _**ALL**_ datasets are now based on `AlbumentationsDataset`.
![graph](https://user-images.githubusercontent.com/13014700/99958765-c56f5f00-2dab-11eb-9308-804505cbf7e9.png)

### DataConfig
`pytorch_learner/learner_config.py`
`pytorch_learner/classification_learner_config.py`
`pytorch_learner/regression_learner_config.py`
`pytorch_learner/object_detection_learner_config.py`
`pytorch_learner/semantic_segmentation_learner_config.py`

To draw a distinction from `GeoData-` classes, the old classes (that work with chips) have been renamed to `ImageData-` where applicable.

### Dataset creation
`pytorch_learner/learner_config.py`
`pytorch_learner/learner.py`

Dataset creation code has been factored out into `Image-` and `Geo-` `DataConfigs` and `Learner`. And removed from the `Learner` subclasses.

### Backend config
`pytorch_backend/pytorch_*_config.py`

To avoid cluttering `PyTorckLearnerBackendConfig`, it now accepts a `DataConfig` directly instead of accepting individual fields of `DataConfig` like it did before. The examples and integration tests have been updated to make use of this.

### Other changes
- ~Adds `persist` param to `RasterioSource` that forces it to keep the file(s) open.~
- Modifies ActivateMixin to make it possible to keep a source permanently activated.
- Adds `lazy` param to `ChipClassificationLabelSource` that makes it skip label reading/inference during initialization.
- Refactors `Box`, `Scene`, `RasterSource`, `SemanticSegmentationLabelSource`, `ChipClassificationLabelSource`, `ChipClassificationLabels` to  make them work better with `GeoDataset`.

### Checklist


## Testing Instructions

* How to test this PR
  - See updated integration tests and examples.
  - ~See new examples (`isprs_potsdam_nochip.py` and `spacenet_rio_nochip.py`).~
  - See updated examples (`isprs_potsdam.py`, `spacenet_rio.py`, and `cowc_potsdam.py`).

Closes #1040
